src/linux.rs

```diff
@@ -0,0 +1,398 @@
+        }
+        if key == "cpu cores" {
+            match value.parse() {
+                Ok(val) => cores = val,
+                Err(_) => break,
+            };
+            chgcount += 1;
+        }
+        if chgcount == 2 {
+            map.insert(physid, cores);
+            chgcount = 0;
+        }
+    }
+    let count = map.into_iter().fold(0, |acc, (_, cores)| acc + cores);
+
+    if count == 0 {
+        get_num_cpus()
+    } else {
+        count
+    }
+}
+
+/// Cached CPUs calculated from cgroups.
+///
+/// If 0, check logical cpus.
+// Allow deprecation warnings, we want to work on older rustc
+#[allow(warnings)]
+static CGROUPS_CPUS: AtomicUsize = ::std::sync::atomic::ATOMIC_USIZE_INIT;
+
+fn cgroups_num_cpus() -> Option<usize> {
+    #[allow(warnings)]
+    static ONCE: Once = ::std::sync::ONCE_INIT;
+
+    ONCE.call_once(init_cgroups);
+
+    let cpus = CGROUPS_CPUS.load(Ordering::Acquire);
+
+    if cpus > 0 {
+        Some(cpus)
+    } else {
+        None
+    }
+}
+
+fn init_cgroups() {
+    // Should only be called once
+    debug_assert!(CGROUPS_CPUS.load(Ordering::SeqCst) == 0);
+
+    match load_cgroups("/proc/self/cgroup", "/proc/self/mountinfo") {
+        Some(quota) => {
+            if quota == 0 {
+                return;
+            }
+
+            let logical = logical_cpus();
+            let count = ::std::cmp::min(quota, logical);
+
+            CGROUPS_CPUS.store(count, Ordering::SeqCst);
+        }
+        None => return,
+    }
+}
+
+fn load_cgroups<P1, P2>(cgroup_proc: P1, mountinfo_proc: P2) -> Option<usize>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+{
+    let subsys = some!(Subsys::load_cpu(cgroup_proc));
+    let mntinfo = some!(MountInfo::load_cpu(mountinfo_proc));
+    let cgroup = some!(Cgroup::mount(mntinfo, subsys));
+    cgroup.cpu_quota()
+}
+
+struct Cgroup {
+    base: PathBuf,
+}
+
+struct MountInfo {
+    root: String,
+    mount_point: String,
+}
+
+struct Subsys {
+    base: String,
+}
+
+impl Cgroup {
+    fn new(dir: PathBuf) -> Cgroup {
+        Cgroup {
+            base: dir,
+        }
+    }
+
+    fn mount(mntinfo: MountInfo, subsys: Subsys) -> Option<Cgroup> {
+        // Translate the subsystem directory via the host paths.
+
+        debug!("subsys = {:?}; root = {:?}; mount_point = {:?}", subsys.base, mntinfo.root, mntinfo.mount_point);
+
+        let rel_from_root = some!(Path::new(&subsys.base).strip_prefix(&mntinfo.root).ok());
+
+        debug!("rel_from_root: {:?}", rel_from_root);
+
+        // join(mp.MountPoint, relPath)
+        let mut path = PathBuf::from(mntinfo.mount_point);
+        path.push(rel_from_root);
+        Some(Cgroup::new(path))
+    }
+
+    fn cpu_quota(&self) -> Option<usize> {
+        let quota_us = some!(self.quota_us());
+        let period_us = some!(self.period_us());
+
+        // protect against dividing by zero
+        if period_us == 0 {
+            return None;
+        }
+
+        Some(quota_us / period_us)
```

Comment:
Good point, JDK does `ceil`. We probably should too, since otherwise a machine with say 8 logical CPUs, but a `1 / 8` would mean we'd see `0`, and default to the logical count of `8`.